File: OkHttpDemo/app/src/main/java/com/example/okhttpdemo/MainActivity.java
package com.example.okhttpdemo;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import com.example.okhttpdemo.okhttp.Call;
import com.example.okhttpdemo.okhttp.CallBack;
import com.example.okhttpdemo.okhttp.OkHttpClient;
import com.example.okhttpdemo.okhttp.Request;
import com.example.okhttpdemo.okhttp.RequestBody;
import com.example.okhttpdemo.okhttp.Response;

import java.io.IOException;

/**
 * @author tengfei
 */
public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity_TAG";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .build();

        RequestBody requestBody = new RequestBody.Builder()
                .type(RequestBody.FROM)
                .addParam("username","hello")
                .addParam("password","123456")
                .build();

        Request request = new Request.Builder()
                .post(requestBody)
                .url("https://www.wanandroid.com/user/login")
                .header("", "")
                .Build();

        Call call = okHttpClient.newCall(request);

        call.enqueue(new CallBack() {
            @Override
            public void onFailure(Call call, IOException e) {
                Log.e(TAG, "onFailure: "+e.getMessage());
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                Log.e(TAG, "onResponse: "+response.toString());
            }
        });
    }
}

File: RetrofitDemo/app/src/main/java/com/example/demo/MainActivity.kt
package com.example.demo

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import okhttp3.*
import java.io.IOException
import android.text.TextUtils
import com.example.demo.okhttp.ExRequestBody
import java.net.URLConnection


private fun guessMimeType(filePath: String): String {
    val fileNameMap = URLConnection.getFileNameMap()

    val mimType = fileNameMap.getContentTypeFor(filePath)

    return if (TextUtils.isEmpty(mimType)) {
        "application/octet-stream"
    } else mimType
}

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val okHttpClient = OkHttpClient.Builder()
            .build()


        val requestBody: RequestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("","")
            .addFormDataPart("file","tengfei.apk", RequestBody.create(MediaType.parse(guessMimeType("")),""))
            .build()

        val exRequestBody:ExRequestBody = ExRequestBody(requestBody)

        val request: Request = Request.Builder()
            .post(exRequestBody)
            .url("")
            .build()

        val call = okHttpClient.newCall(request)

        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

            override fun onResponse(call: Call, response: Response) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

        })

    }
}
